### ISSUE TITLE
Not able to Receive or send data

### ISSUE THREAD

**Issue by esa1966** [abouillot/HomeAutomation#6]
Hi,
I'm not able to send or receive data with the RPI Gateway.
I don't know how to debug the system.
A little advice would be fine.

I have done the following :
1) Install Mosquitto 1.4 on the PI.
When I run it it seems fine :

root@(none):/etc/mosquitto# mosquitto -c /etc/mosquitto/mosquitto.conf
1426432545: mosquitto version 1.4 (build date 2015-03-15 11:02:37+0000) starting
1426432545: Config loaded from /etc/mosquitto/mosquitto.conf.
1426432545: Opening ipv4 listen socket on port 1883.
1426432545: Opening ipv6 listen socket on port 1883.
1426432545: Warning: Address family not supported by protocol
1426432545: Opening websockets listen socket on port 9001.
1426432545: Warning: Mosquitto should not be run as root/administrator
2) installed and ran the gateway : HomeAutomation

insert code hereroot@(none):/home/HomeAutomation/piGateway# sudo ./Gateway
sudo: unable to resolve host new-host

Listening at 433 Mhz...setup complete
Connect succeed
Must I see something here when I receive data from a node?

3) I use a node to transmit data periodically: https://github.com/homeautomationyay/OpenHab-RFM69/blob/master/UberSensor.ino

I check the node with a gateway sketch from LowPowerLab (I downloaded it on an Anarduino) and I received the data string.
But, If I set the same adress on the Rpi gateway (I power of the Arduino Gateway) I don't receive data.
I also try with another "test gateway" found from "http://rdepablos.merlitec.com/mixed/rfm69-library-for-raspberry-pi" and I receive nothing.

4) No Data Send with PiGateway
Then, I modified the gateway as a node adress 11 to send data to the Arduino Gateway.
But, I don't receive data on the gateway.

I have done the following test :

```
SPI : tested.
hardware Connection between RPi and RFM69: tested.
the NetworkID : verified.
the crypkey : verified
```

Where is my mistake?
Question :

```
How to debug the RFM69CW part of the gatewayPi?
How can a check the SPI connection with the RFM69? And what data should I receive?
```

Thank

**Comment by abouillot** [comment]
Have you check the network id and the encrypt key to be the same accross
the pi and node?

If the connect succeed is displayed, the communication between the pi and
the RFM is running.

**Comment by mccare** [comment]
Did you manage to fix this? I'm also currently stuck in not being able to receive data on the RPI coming from an Arduino using the RFM69CW chip.

**Comment by abouillot** [comment]
I haven't heard from Esa since its post. Is the PI communicating properly with the RFM?
Do you see : 
Listening at 433 Mhz...setup complete
Connect succeed
in the log or on the screen?

**Comment by mccare** [comment]
yes. I will try again tomorrow with 2 rasperry pis. my sending arduino is 5V. I think the RFM69 doesn't like 5V input pins... I have RFM69CW, but hope they work the same as the RFM69W

**Comment by mccare** [comment]
So, more info:

I see the info "Listening at 868Mhz" and then the setup complete. I've tried to set up two PIs with RFM69CW connected and tried to get them to talk to each other. (simplified the Gateway.c to just send and receive packets, see https://github.com/mccare/HomeAutomation/blob/send_receive/piGateway/Gateway.c)

Anyway, no luck there. I think I've attached it correctly but either
- the chip is damaged since I've used it with 5V input for MISO/MOSI/NSS/CLK/DI0
- the RFM69CW is different than the RFM69W (but the lowpowerlab library states that it is supported).

The Gateway will tell me that it sends data, but receives no ACK, the other gateway says the same. I assume if they are just 1m appart I don't need to solder an antenna to them? (I tried a 82mm long copper wire)

I have ordered some RFM69W and will try with them.

Christian

**Comment by abouillot** [comment]
As the setup goes to its end, I don't think the RFM are not fully burned, even if they are supposed to work at 3.3V max, as the SPI is working and the RFM respond to the config commands. I'm not sure about the DIO0 - the interrupt line - as it is not used in setup part. 
I suggest you use some level shifter between the Arduino and the RFM, or source some 3.3V Arduino.

I've run your version on two raspberry connected to RFM69HW in 433MHz and had it running.

**Comment by esa1966** [comment]
Hi,
I checked your version but only on one Rpi. The other side is a node or the GW on a Arnarduino.

1) With anarduino => anarduino Node 10 transmit to anarduino GW 01 => OK
2) I stop anarduino GW01 and launch GW Rpi => no reception from anarduino node 10
3) I stop anarduino Node 10 and Start Node Rpi and start GW 01 anarduino => no reception

When I disconnect the RFM69 from the Pi connector then the program stop (freeze) with only one line printed. I suppose the RFM is correctly detected and SPI are OK?

Remark : I have a Rpi B  512 Mo

This is the info about SPI : 
ls -al /dev/spi*
crw------- 1 root root 153, 0 Jan  1  1970 /dev/spidev0.0
crw------- 1 root root 153, 1 Jan  1  1970 /dev/spidev0.1

lsmod | grep spi
spi_bcm2835             7948  0

dmesg|grep spi
[    4.674552] spi spi0.0: setting up native-CS0 as GPIO 8
[    4.692302] spi spi0.1: setting up native-CS1 as GPIO 7

In raspi-config I get this error when I enable the SPI interface
modprobe: ERROR: could not insert 'spi_bcm2708': No such device

Is it Normal?
Thank

**Comment by mccare** [comment]
I get the same error as you from raspi-config when enabling SPI. Since I now have a working send/receive example running, I don't think this is the problem. 

The solution to my problem was:
- use shorter wire (less than 5cm instead of 10-15cm)
- dial down the SPI_SPEED (see below) to 500khz

in rfm69.h

```
#define SPI_SPEED 500000
//#define SPI_SPEED 2000000
```

I'll try to clean up my code base to get a simple test case with send/receive for future debugging :)

**Comment by abouillot** [comment]
I'm glad you figured out how to address your issue. I integrated the send/receive code so everybody will be able to enjoy it when needed to check comms.

For your information, in my setup I used 20 cm Dupont cable from the Raspberry to a bread board holding the RFM. I didn't had issue with the SPI for the last year or so getting data on the gateway.

**Comment by esa1966** [comment]
Dear,
I have followed your recommendation, but it doesn't solve my problem.

To be sure that I didn't make an error in the source file, could you try on your system?
Bellow, the node running on the Anarduino (at 433Mhz) : 
[NodeAbuillot.txt](https://github.com/abouillot/HomeAutomation/files/51206/NodeAbuillot.txt)

The Gateway (rename extension txt to zip)
[piGateway.txt](https://github.com/abouillot/HomeAutomation/files/51208/piGateway.txt)

If it run on your setup, then my Rpi is maybe defect or a problem of wiring.

Thank for your help

**Comment by mccare** [comment]
I don't have an Arduino client yet. To debug further, on your Pi, please go to the rfm69.cpp into the RFM69::interruptHandler routine and uncomment the print statement. If the chip has received something, it will signal this by setting the DIO0 to high and this will trigger the interrupt. So if you don't see this print statement there, the chip didn't receive a packet.

Also check that the line telling you the network and node id is correct (and that what you expect). You might also be able to enable promiscousMode. 

For further debuggin I have put further print statements into the interrupt handler routine (where it tries to read something).

**Comment by esa1966** [comment]
Hi,

Thank a lot for your help.

I made the following test :

I have an Anarduino node that send a string periodically : node 10 to GW1 - NetworkID 100 - Freq 433 - Key = "sampleEncryptKey"

1) I place a scope on the DIO0 pin of the RFM69 on the Anarduino GW.
I can see a pulse each time it receives the string.

2) I place a scope on the DIO0 pin of the RFM69 on the Rpi.
I don't see a pulse.

3) I run "SenderReceiver -s" (the simple test made by Abouillot).
When the Rpi send the string then I can see a pulse on the DIO0 pin.
It's like the RFM69 receive its own packet.
So, I suppose the RFM69 module is working.

It seems that the problem come from a compatibility issue between the RFM69 wired on the Anarduino and the RFM69 wired on the Rpi.
I checked that the two modules are the same frequency (I can't see the inscription on the RFM69 wired on the Anarduino, but I ordered a 433Mhz and I run the program at 433Mhz).
It's strange.

**Comment by esa1966** [comment]
Hi,
I finally have a reception running "SenderReceiver" on the two Rpi.

By running successively "SenderReceiver -s" on the two Rpi, I'm able to see a reception on Rpi1 and on Rpi2 : that's fine. The two RFM 69 modules are working.

Thanks a lot to have made this program and giving me the possibility to check the setup with two Rpi!

But I only have a reception every +/- 5 transmissions.
This is exactly the same behaviour on the two Rpi.
I can imagine that one module might be defect (in reception) but not the two modules.
Question 1 : Do you have the same behaviour?

More info this morning : I check the Interrupt pulse (0.5ms) and noticed that I didn't receive it every send message and I suppose the packet is lost.
But, I count the pulses on my scope and on 10 pulses, the program triggered only one receive.

Question 2 : why it doesn't receive when I transmit with the Anarduino.
could you check the code (I take it from the folder /HomeAutomation/SensorNode/SensorNode.ino.)
I just change the frequency to 433Mhz.

Thank for your help

**Comment by mccare** [comment]
No, I do have reception on every packet. Did you solder an antenna to your chip? A simple wire with a certain length (depending on your frequency) should be enough. I haven't had a chance to setup my arduinos yet. (just arrived today).

**Comment by esa1966** [comment]
Yes I have antenna (same size than for my arduino that work without missing packet)
The problem is that it receives something because I see a pulse on DIO0 but It doesn't trigger the interrupt.

> I check the Interrupt pulse (0.5ms) and noticed that I didn't receive it every send message and I suppose the packet is lost.
> But, I count the pulses on my scope and on 10 pulses, the program triggered only one receive.

**Comment by mccare** [comment]
I assume you have the newest wiringPi library, the current jessie debian OS, update firmware and the like? 

I have just tested the master branch from abouillots version on my system. My system has problems with the SPI_SPEED that is in rfm69.h. If I dial it down to 500000 (500Khz) a test works, otherwise just intermittent. 

My system:
Linux sensornode 4.1.13+ #826 PREEMPT Fri Nov 13 20:13:22 GMT 2015 armv6l GNU/Linux

**Comment by esa1966** [comment]
Hi,
You are right! 
SPI at 500khz is the solution and in this case, I don't lose packet (more precisely, I add a counter and I only lose 1 packet on 200).
Thank

I'm interested by your test with your Arduino modules.
As explained, I took the example code SensorNode.ino in the folder HomeAutomation/SensorNode/.
I suppose it must work with this code (I only adapted the frequency)?

**Comment by mccare** [comment]
I tested my Arduino model and it worked with the SensorNode code. I've removed all the DHT22 stuff since I need to dimm an LED driver through PWM signal. But sending commands to the Pi gateway worked fine. I noticed that I had the isRFM69HW high flag set in the Gateway.c code. That produced unexpected results (the ACK was not delivered from the gateway but the gateway was able to receive data).

**Comment by esa1966** [comment]
Hi,
I'm glad to ear that it should work.
In my case, I made the following modification : 
#define NETWORKID     100
#define GATEWAYID     10
#define ENCRYPTKEY    "sampleEncryptKey" //exactly the same 16 characters/bytes on all nodes!
//#define IS_RFM69HW    //uncomment only for RFM69HW! Leave out if you have RFM69W!
Suppress  reference to DHT : 
//#include <dht.h>
//dht DHT;
/*int chk = DHT.read21(DHTPIN);

  switch (chk)
    {
    case DHTLIB_OK:
        DEBUG1("OK,\t");
        break;
    case DHTLIB_ERROR_CHECKSUM:
        DEBUG1("Checksum error,\t");
        break;
    case DHTLIB_ERROR_TIMEOUT:
        DEBUG1("Time out error,\t");
        break;
    case DHTLIB_ERROR_CONNECT:
        Serial.print("Connect error,\t");
        break;
    case DHTLIB_ERROR_ACK_L:
        DEBUG1("Ack Low error,\t");
        break;
    case DHTLIB_ERROR_ACK_H:
        DEBUG1("Ack High error,\t");
        break;
    default:
        DEBUG1("Unknown error,\t");
        break;
    }
*/

Set two values : 
double h = 55;
    // Read temperature as Celsius
    double t = 22;

I run "./SenderReceiver -r" on the Rpi but I don't receive something.
My code below : 
[NodeAbuillot.txt](https://github.com/abouillot/HomeAutomation/files/57692/NodeAbuillot.txt)

Can you check on your arduino?

**Comment by abouillot** [comment]
I've checked the node code you referenced (NodeAbouillot.txt), with the change as RFM69_HW, as this is my setup, with the SenderReceiver client on the Pi and got it run. The amount of message correctly received depend on the SPI speed and Pi overclocking. But always get some messages whatever settings.
you mention "./SenderReceiver -r" as command line, but you need to be root so it should read "sudo ./SenderReceiver -r", but I gess you've done so, as a warning is displayed and the program wouldn't run.

**Comment by esa1966** [comment]
Hi Alexandre, Christian,
Thank a lot to have checked the code. It help me to fix the second issue.
As the code is fine, I checked the RFM69 registers and have add the function readAllRegs() to the code. 

> **Remark** : To have the same printout with binary value (as for Arduino) I modified the code. Maybe you can add it :
>  void RFM69::readAllRegs()
> {
> #ifdef RASPBERRY
>   char thedata[2];
>   int i;
>   thedata[1] = 0;
> const char *bit_rep[16] = {
>     [ 0] = "0000", [ 1] = "0001", [ 2] = "0010", [ 3] = "0011",
>     [ 4] = "0100", [ 5] = "0101", [ 6] = "0110", [ 7] = "0111",
>     [ 8] = "1000", [ 9] = "1001", [10] = "1010", [11] = "1011",
>     [12] = "1100", [13] = "1101", [14] = "1110", [15] = "1111",
> };
>   for(i = 1; i <= 0x4F; i++) {
>    printf("%i - %x - ", i, readReg(i));
>    printf("%s%s \n\r", bit_rep[readReg(i) >> 4], bit_rep[readReg(i) & 0x0F]);
>   }
> #else
> ...

Doing this I saw that the Bitrate (register 3&4) and FDev (register 5&6) wasn't the same on the Rpi and the Arduino.

> **This was from arduino**
> Node Arduino : Bit rate (0x2E66) = 2.7 kb/s &  Fdev (0x5C3) = 1475 x 61 = 89975 kHz
> 1 - 4 - 100
> 2 - 0 - 0
> 3 - 2E - 101110
> 4 - 66 - 1100110
> 5 - 5 - 101
> 6 - C3 - 11000011
> 7 - 6C - 1101100
> 8 - 40 - 1000000
> 9 - 0 - 0
> 
> **From the  Rpi** : Bit rate (0x240) = 55 kb/s &  Fdev (0x333) = 819 x 61 = 49959 kHz
> 1 - 4 - 00000100
> 2 - 0 - 00000000
> 3 - 2 - 00000010
> 4 - 40 - 01000000
> 5 - 3 - 00000011
> 6 - 33 - 00110011
> 7 - 6c - 01101100
> 8 - 40 - 01000000
> 9 - 0 - 00000000

I realised that the Library used for the Arduino wasn't the Moteino library but another one I got on Ethernet a long time ago. Sorry, my fault :-(

Everything work fine now. I can start a Node to send value and the two Gateway (one on the RPi and the other on a Arduino). I can now receive the string on the two Gateway.

To summarize the issue  : 
- For the Rpi Gateway : the SPI Speed must be set to 500 Khz.
- For the Arduino, I have to take the last library from Moteino GitHub (https://github.com/LowPowerLab/RFM69)

Thank for your help.

**Comment by abouillot** [comment]
Glad you succeed getting your setup up and running. The rfm69.cpp can be integrated for Arduino, as the porting want to be transparent. This way you main on-par on Arduino and Pi side.
